core/src/nodes/handled_node_tasks.rs

```diff
@@ -450,3 +451,380 @@ where
+                        future: self.inner_fut.take().unwrap(),
+                        handler: Handler::default(),
+                        events_buffer: Vec::new(),
+                    }
+                };
+                let node_task = NodeTask {
+                    inner: inner,
+                    events_tx: events_from_node_task_tx.clone(), // events TO the outside
+                    in_events_rx: events_to_node_task_rx.fuse(), // events FROM the outside
+                    id: self.task_id,
+                };
+                (node_task, events_to_node_task_tx, events_from_node_task_rx)
+            }
+        }
+
+        #[test]
+        fn task_emits_event_when_things_happen_in_the_node() {
+            let (node_task, tx, mut rx) = TestBuilder::new()
+                .with_task_id(890)
+                .node_task();
+
+            tx.unbounded_send(InEvent::Custom("beef")).expect("send to NodeTask should work");
+            let mut rt = Runtime::new().unwrap();
+            rt.spawn(node_task);
+            let events = rt.block_on(rx.by_ref().take(2).collect()).expect("reading on rx should work");
+
+            assert_matches!(events[0], (InToExtMessage::NodeReached(_), TaskId(890)));
+            assert_matches!(events[1], (InToExtMessage::NodeEvent(ref outevent), TaskId(890)) => {
+                assert_matches!(outevent, OutEvent::Custom(beef) => {
+                    assert_eq!(beef, &"beef");
+                })
+            });
+        }
+
+        #[test]
+        fn task_exits_when_node_errors() {
+            let mut rt = Runtime::new().unwrap();
+            let (node_task, _tx, rx) = TestBuilder::new()
+                .with_inner_fut(future::err(io::Error::new(io::ErrorKind::Other, "nah")))
+                .with_task_id(345)
+                .node_task();
+
+            rt.spawn(node_task);
+            let events = rt.block_on(rx.collect()).expect("rx failed");
+            assert!(events.len() == 1);
+            assert_matches!(events[0], (InToExtMessage::TaskClosed{..}, TaskId(345)));
+        }
+
+        #[test]
+        fn task_exits_when_node_is_done() {
+            let mut rt = Runtime::new().unwrap();
+            let fut = {
+                let peer_id = PublicKey::Rsa((0 .. 2048).map(|_| -> u8 { random() }).collect()).into_peer_id();
+                let mut muxer = DummyMuxer::new();
+                muxer.set_inbound_connection_state(DummyConnectionState::Closed);
+                muxer.set_outbound_connection_state(DummyConnectionState::Closed);
+                future::ok((peer_id, muxer))
+            };
+            let (node_task, tx, rx) = TestBuilder::new()
+                .with_inner_fut(fut)
+                .with_task_id(345)
+                .node_task();
+
+            // Even though we set up the muxer outbound state to be `Closed` we
+            // still need to open a substream or the outbound state will never
+            // be checked.
```

Comment:
This should be fixed in the non-test code.

Reply:
I'll look into it, but I suspect that it's more of an artifact of the `DummyMuxer` than a fault of the actual code.

Reply:
I am wrong and you're right (and the comment is correct); the `NodeStream` does not check if the outbound state is open or closed (it does check the inbound state).

I'll open an issue and fix in separate PR, ok?

Reply:
Issue https://github.com/libp2p/rust-libp2p/issues/609